dscache/dscache.go

```diff
@@ -170,52 +174,110 @@ func (d *Dscache) ListRefs() ([]reporef.DatasetRef, error) {
 	return refs, nil
 }
 
-// Update modifies the dscache according to the provided action.
-func (d *Dscache) Update(act *logbook.Action) error {
+// Subscribe watches a logbook for actions that change the state
+func (d *Dscache) Subscribe(bus event.Bus) {
+	d.Messages = bus.Subscribe(event.ETDatasetInit, event.ETDatasetChange)
+	go func() {
+		for {
+			e := <-d.Messages
+			if dsChange, ok := e.Payload.(event.DatasetChangeEvent); ok {
+				if e.Topic == event.ETDatasetInit {
+					if err := d.updateInitDataset(dsChange); err != nil {
+						log.Error(err)
+					}
+				} else if e.Topic == event.ETDatasetChange {
+					if err := d.updateMoveCursor(dsChange); err != nil {
+						log.Error(err)
+					}
+				}
+			}
+		}
+	}()
+}
```

Comment:
Ok lots to talk about here. Here's how I think we should refactor:

```go

// We need to not leak goroutines. The context passed into the `New` constructor
// needs to be plumbed into here and cleaned up if the passed-in context is 
// cancelled.
func (d *Dscache) subscribe(ctx context.Context, bus event.Bus) {
  // Subsystems that are aware of events should not have an event channel field.
  // This allows code that breaks the encapsulation of the `subscribe` method. 
  // Instead `subscribe` should use a local variable.
  eventsCh := bus.Subscribe(event.ETDatasetInit, event.ETDatasetChange)
  go func() {
    for {
      select {
        // The statement that listens for events should use the "commma ok"
        // pattern to check for a closed channel. If the event bus context is
        // cancelled before this goroutine exits, the event bus should close
        // listener channels. Breaking on !ok will gracefully close
        case e, ok := <-eventsCh:
          if !ok {
            break
          }
          if dsChange, ok := e.Payload.(event.DatasetChangeEvent); ok {
            if e.Topic == event.ETDatasetInit {
              if err := d.updateInitDataset(dsChange); err != nil {
                log.Error(err)
              }
            } else if e.Topic == event.ETDatasetChange {
              if err := d.updateMoveCursor(dsChange); err != nil {
                log.Error(err)
              }
            }
          }
        case <-ctx.Done():
          // TODO (b5): add unsubscribe method to bus interface, the arg for
          // Unsubscribe should be the channel variable itself, which will be 
          // available in this scope:
          // bus.Unsubscribe(eventsCh)
          break
      }
    }()
  }
}
```

The general pattern I like to see emerge is a private `subscribe` method that encapsulates all event handling for a subsystem. We should use the same private method name across _all_ subsystems for clarity. `subscribe` should _always_ be called within the constructor and listen to the context given to the constructor. Passing a non-nil event bus to a constructor should _always_ subscribe, cutting down on potential bugs caused by requiring package consumers to both construct & subscribe.


Finally, I think this is a good enough reason to add an `Unsubscribe` method to the bus interface, and call it if the context is cancelled. Currently we can get away with this because in all (production) cases the context given to `Dscache` will live for the same duration as the bus itself. But this might not always be the case (I'm thinking of other subsystems like `cron`). Having all subscribers clean up on close means we won't have to rely on assumption of  aligned context duration. I'd be happy to PR in the Unsubscribe method.

Reply:
This all makes sense, added as described